Fix subnet unsubscription time

pawanjay176 opened the pull request:
## Issue Addressed

Hopefully fixes https://github.com/sigp/lighthouse/issues/6732

## Proposed Changes

In our `scheduled_subscriptions`, we were setting unsubscription slot to be `current_slot + 1`. Given that we were subscribing to the subnet at `duty.slot - 1`, the unsubscription slot ended up being `duty.slot`. So we were unsubscribing to the subnet at the beginning of the duty slot which is insane.

Fixes the `scheduled_subscriptions` to unsubscribe at `duty.slot + 1`.

## Additional Info

Running this on kurtosis made me realise that we do not send subscriptions for the first couple of slots because of https://github.com/sigp/lighthouse/blob/unstable/validator_client/validator_services/src/duties_service.rs#L79-L85

Not sure if we want to handle that.

pawanjay176 committed:
Fix the end_slot in delayed subscriptions

michaelsproul reviewed (approved):
Nice find!

The fix makes sense to me.

The testnet case is interesting, but I think a fix for that can wait for another PR. I guess we might want to special-case the VC's/BN's handling of the first ~2 slots of a network (feels kind of dumb), _or_ remove the loud warnings for late subscriptions from the BN and just track them with a metric.

michaelsproul commented:
Looks like one of the tests is failing. It might actually be expecting the old behaviour :eyes:

pawanjay176 committed:
Fix test

pawanjay176 commented:
I fixed the failing test according to my understanding.
I think prior to #6146 , the test was checking if we bundled multiple unsubscriptions together into a single one (e.g. if we need to unsubscribe from subnet X at slot 9 and 12, then skip the unsubscribe at 9).

After the refactor, we decide to extend the unsubscription only in `subscribe_to_subnet_immediately` 1 slot before the subscription. So multiple unsubscriptions will get bundled together only in the case where there's overlap between an existing subscription and the new one (e.g. if we are subscribing at slot 11 and there's a planned unsubscription event at slot 12, then we just extend the subscription). 

TL;DR: the test `test_subscribe_same_subnet_several_slots_apart` wasn't refactored with the logic change in #6146 , and started failing when we fixed a bug. cbf7dca refactors the test.

jimmygchen reviewed (commented):
Nice find! 👍

dapplion commented on beacon_node/network/src/subnet_service/tests/mod.rs:
```diff
@@ -119,7 +118,7 @@ fn get_logger(log_level: Option<slog::Level>) -> Logger {
 
 fn get_subnet_service() -> SubnetService<TestBeaconChainType> {
-    let log = test_logger();
+    let log = get_logger(TEST_LOG_LEVEL);
```
Why not use the test_logger? We want to toggle logs programmatically with `--features test_logger`

pawanjay176 replied:
I set it to be consistent with `TestBeaconChain` since it uses `get_logger(TEST_LOG_LEVEL)`. 
Didn't know the `test_logger` feature was a thing. I'll change both to `test_logger()`

dapplion commented on beacon_node/network/src/subnet_service/tests/mod.rs:
```diff
@@ -607,9 +604,18 @@ mod test {
         assert_eq!(no_events, []);
 
-        let second_subscribe_event = get_events(&mut subnet_service, None, 2).await;
+        let subscription_end_slot = current_slot + subscription_slot2 + 2; // +1 to get to the end of the duty slot, +1 for the slot to complete
```
Place the comments above

dapplion commented on beacon_node/network/src/subnet_service/mod.rs:
```diff
@@ -629,9 +635,8 @@ impl<T: BeaconChainTypes> Stream for SubnetService<T> {
-                let current_slot = self.beacon_chain.slot_clock.now().unwrap_or_default();
-                if let Err(e) = self.subscribe_to_subnet_immediately(subnet, current_slot + 1) {
+                let ExactSubnet { subnet, slot } = exact_subnet;
+                if let Err(e) = self.subscribe_to_subnet_immediately(subnet, slot + 1) {
```
You should add some comments referencing the issue and the reason for `slot + 1` in the code. Useful if we ever refactor this and git blame breaks

pawanjay176 committed:
Address lion's comments

pawanjay176 committed:
fmt and lint

dapplion reviewed (approved):
LGTM

michaelsproul commented:
Fix for CI failure:

- https://github.com/sigp/lighthouse/pull/6901

AgeManning reviewed (approved):
Yeah nice catch. 

I wouldn't have imagined this would change the density of subscribed subnets between 5.3 and 6.0?

jimmygchen commented:
@mergify requeue

jimmygchen commented:
> Yeah nice catch.
> 
> I wouldn't have imagined this would change the density of subscribed subnets between 5.3 and 6.0?

Looks like this was introduced in #6146 which was part of the 6.0 release (My bad, I missed this in the review!). 

The old logic was good and same as the fix in this PR:
https://github.com/sigp/lighthouse/blob/08e8b92e5032044f253a7357aaf0a7c74f9f8b80/beacon_node/network/src/subnet_service/attestation_subnets.rs#L644-L647
